agent/src/beerocks/slave/tasks/task.h

```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TASK_H_
+#define _TASK_H_
+
+#include <beerocks/tlvf/beerocks_header.h>
+#include <tlvf/CmduMessageRx.h>
+
+namespace beerocks {
+
+class Task {
+public:
+    Task() {}
+    virtual ~Task() {}
+
+    /**
+     * @brief Enum which defines all tasks type.
+     * 
+     * Each task must have an entry on the enum eType, so it will be possible to send the task an
+     * event. 
+     */
+    enum eType : uint8_t {
+        TOPOLOGY,
+    };
+
+    /**
+     * @brief Retrun the task type.
+     * 
+     * Called by TaskPool send_event().
+     * 
+     * @return eType Task type.
+     */
+    virtual eType get_task_type() = 0;
+
+    /**
+     * @brief Work function. Shall do the task routine. Called by TaskPool run_task() function.
+     */
+    virtual void work(){};
+
+    /**
+     * @brief Handle events, which are being send from containing thread.
+     * 
+     * @param event_enum_value Event enum value which shall be defined on the task.
+     */
+    virtual void handle_event(uint8_t event_enum_value){};
+
+    /**
+     * @brief Handle CMDU message.
+     * 
+     * @param cmdu_rx CMDU object conataining the message.
+     * @param src_mac AL Mac of the sender.
+     * @param beerocks_header Beerocks Message header (Only on VS message).
+     * @return true if the message has been handled, otherwise false.
+     */
+    virtual bool handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac,
```

Comment:
For what? So I would have to create a struct, copy to it all data members before passing it to the handler?
I don't agree with it.

Reply:
For many reasons:
>   setting and validation of the internal relationships between the various
>   fields that are required to handle cmduRx.
>    e.g. is it legal to have beerocks_header == nullptr?
>    e.g. can we work with ZERO_MAC as the src addr? does it imply somethimg?
>    etc.

The alternative is to leave these validations and integrity of the data to the writer of the function. In this case either:
* it wouldn't be done.
* it would be done wrongly
* it would be re-written (copy/pasted) along the code.

Reply:
It is legal to have beerocks_header=nullptr.
And it is legal that src_mac=ZERO_MAC.

Reply:
You are completely missing the point.
Once you have a structure (class) that represents CMDU to be handled you _have a place to enforce, to validate, to check_ the relation between these parameters.
Even if now you don't see it, I am quite sure that there _are_ relationships that we would like to enforce, now, or in the future. Spreading these 3 parameters in each function call just makes it much harder to enforce them now or in the future.
And don't worry about copy - this structure has only references and pointers to existing objects. it is cheep - i.e. you will not pay more than you pay now when you pass the variables separately.